src/integrators/aov.cpp

```diff
@@ -356,12 +425,98 @@ class AOVIntegrator final : public SamplingIntegrator<Float, Spectrum> {
+
+    TensorXf get_channels_slice(const TensorXf& src, size_t channel_offset, size_t num_channels) const {
+        using Array = typename TensorXf::Array;
+
+        size_t slice_shape[] = { src.shape(0), src.shape(1), num_channels };
+        uint32_t slice_flat = slice_shape[0] * slice_shape[1] * slice_shape[2];
+
+        DynamicBuffer<UInt32> idx = dr::arange<DynamicBuffer<UInt32>>(slice_flat);
+        DynamicBuffer<UInt32> pixel_idx = idx / num_channels;
+        DynamicBuffer<UInt32> channel_idx = dr::fmadd(pixel_idx, uint32_t(-(int)num_channels), idx) 
+            + channel_offset;
+
+        auto values_idx = dr::fmadd(pixel_idx, src.shape(2), channel_idx);
+        return TensorXf(dr::gather<Array>(src.array(), values_idx), 3, slice_shape);
+    }
+
+    void set_channels_slice(const TensorXf& src, TensorXf& dst, size_t dst_channel_offset) {
+        auto* src_shape = src.shape().data();
+        uint32_t src_flat = src_shape[0] * src_shape[1] * src_shape[2];
+
+        DynamicBuffer<UInt32> idx = dr::arange<DynamicBuffer<UInt32>>(src_flat);
+        DynamicBuffer<UInt32> pixel_idx = idx / src_shape[2];
+        DynamicBuffer<UInt32> dst_channel_idx = dr::fmadd(pixel_idx, uint32_t(-(int)src_shape[2]), idx) 
+            + dst_channel_offset;
+
+        uint32_t num_dst_channels = dst.shape(2);
+        auto dst_values_idx = dr::fmadd(pixel_idx, num_dst_channels, dst_channel_idx);
+
+        dr::scatter(
+            dst.array(),
+            src.array(),
+            dst_values_idx);
+    }
+
+    /// Combine inner integrator images and AOVS image
+    TensorXf merge_channels(const std::vector<TensorXf>& inner_images, 
+                          const TensorXf& aovs_image) {
+        using Array = typename TensorXf::Array;
+
+        size_t num_aovs = m_aov_names.size();
+
+        auto* aovs_image_shape = aovs_image.shape().data();
+
+        // Figure out entire number of channels of combined image
+        size_t combined_shape[3] = { aovs_image_shape[0], aovs_image_shape[1], num_aovs };
+        for (const auto& image : inner_images)
+            combined_shape[2] += image.shape(2);
+
+        size_t combined_flat = combined_shape[0] * combined_shape[1] * combined_shape[2];
+        TensorXf combined_image = TensorXf(dr::zeros<Array>(combined_flat), 3, combined_shape);
+
+        // Load base channels from inner integrators into combined tensor
+        uint32_t channel_offset = 0;
+        for (const auto& image : inner_images) {
+            set_channels_slice(image, combined_image, channel_offset);
+            channel_offset += image.shape(2);
+        }
+
+        // Load aovs image into combined
+        set_channels_slice(aovs_image, combined_image, channel_offset);
+
+        return combined_image;
+    }
+
+    /// Split up an image into image generated by inner-integrators and AOV image
+    std::pair<std::vector<TensorXf>, TensorXf> split_channels(size_t base_channel_count, const TensorXf& combined_image) const {
```

Comment:
Should those methods be placed in a different file, accessible to the rest of the codebase? Looks pretty generic functionalities to me.